microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/utils/CompletableFutureWrapper.java

```diff
@@ -0,0 +1,294 @@
+/*******************************************************************************
+* Copyright (c) 2023 Red Hat Inc. and others.
+*
+* This program and the accompanying materials are made available under the
+* terms of the Eclipse Public License v. 2.0 which is available at
+* http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+* which is available at https://www.apache.org/licenses/LICENSE-2.0.
+*
+* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lsp4mp.utils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Wrapper of a given {@link CompletableFuture} to add some
+ * {@link CompletableFuture} which need to be canceled when the wrapped
+ * {@link CompletableFuture} is cancelled.
+ * 
+ * @author Angelo ZERR
+ *
+ * @param <T>
+ */
+class CompletableFutureWrapper<T> extends CompletableFuture<T> implements ExtendedCancelChecker {
+
+	private final CompletableFuture<T> delegate;
+	private List<CompletableFuture<?>> futuresToCancel;
+
+	public CompletableFutureWrapper(CompletableFuture<T> delegate) {
+		this.delegate = delegate;
+		this.futuresToCancel = new ArrayList<>();
+	}
+
+	@Override
+	public void checkCanceled() {
+		if (delegate.isCancelled())
+			throw new CancellationException();
+	}
+
+	@Override
+	public <T> CompletableFuture<T> cancelIfNeeded(CompletableFuture<T> futureToCancel) {
+		this.futuresToCancel.add(futureToCancel);
+		return futureToCancel;
+	}
+
+	/**
+	 * Cancel the delegate future and the list of future to cancel.
+	 */
+	public boolean cancel(boolean mayInterruptIfRunning) {

```

Comment:
I think it would be nice to add the `@Override` to all the methods. There should be a clean up to do this quickly. If you don't think it's that important, feel free to ignore this comment.